 #! /usr/bin/env python
# -*- coding: utf-8 -*-

# Setup
from __future__ import (absolute_import, division,
                        print_function, unicode_literals)
from builtins import *

import re
import gzip
import pandas as pd
from collections import defaultdict

from snakemake.utils import min_version

min_version("6.4.1")

from snakemake.exceptions import RuleException

# config file
configfile: "config/config.yaml"

# read single cell samples table
sc_subsamples = pd.read_table('config/pep/snRNAseq_GTEX_stellarscope_subsample_table.csv', sep=',')

# read samples table
gtex_samples = pd.read_table(config["samples_tsv"]).set_index("tissue_id", drop=False)

# params
# We are not generating multiple reassignment modes for now
#s_reassignment = ["all", "average", "choose", "conf", "exclude", "unique"]

rule all:
    input:
        expand("results/completed/{s}_bulk_completed.txt", s=gtex_samples["bulk_RNAseq"]),
#        expand("results/completed/{s}_sc_{s_method}_{reassignment}_completed.txt", s=gtex_samples["sn_RNAseq"], s_method=["individual", "pseudobulk"], reassignment = s_reassignment)
        expand("results/completed/{s}_sc_{s_method}_rep{copies}_{strands}_completed.txt", s=gtex_samples["sn_RNAseq"], s_method=["pseudobulk"], copies=[1,2], strands=["U","F"])

# Rules to include
#include: "rules/gdc_download.smk"
include: "rules/references.smk"
include: "rules/bam_to_fastq.smk"
include: "rules/star_index.smk"
include: "rules/star_alignment.smk"
include: "rules/telescope.smk"
include: "rules/make_sc_sample_fastqs.smk"
include: "rules/starsolo_alignment.smk"
include: "rules/stellarscope_cellsort.smk"
include: "rules/stellarscope_13.smk"
#include: "rules/stellarscope.smk"
#include: "rules/autoread_seurat.smk"

